### PR TITLE
Fix simplify-exits

### PR DESCRIPTION
**Pull request by Ekdohibs**
Depends on #1107.

### CODE REVIEW

**Comment by mshinwell** [comment]
@chambart to review (don't merge until #1107 merged please)

**Comment by Ekdohibs** [comment]
Rebased on top of #1107, as it would be good to merge those immediately after one another.

File: ocaml/lambda/simplif.ml
Comment (chambart) [inline ocaml/lambda/simplif.ml:227]: I think that it is possible to replace by

```suggestion
    let result_layout ly =
      match layout with
      | None -> ly
      | Some layout ->
          assert(Typeopt.compatible_layout layout ly);
          layout
    in
```

I think that if all layout are correctly generated from the type, given that staticraise is not part of the source language, all the expressions replaced by a staticraise must have the correct type.

**Review by chambart — APPROVED** [review]
I think that this is ok.

File: ocaml/lambda/simplif.ml
Comment (Ekdohibs) [reply ocaml/lambda/simplif.ml:227]: Unfortunately, I think this will not work, at least without substantial changes to the producers of code, such as https://github.com/ocaml-flambda/flambda-backend/blob/2142e4c1d8e6d24b83d84e3684bcdeb63ed197a2/ocaml/lambda/matching.ml#L3584 .
Here, the `map_return` function keeps the existing layouts, even if the return value gets transformed to a `Lstaticraise`. It is probably possible to change it to fix this problem, but it is unclear if it is worth it compared to the additional complexity in all producers, as well as adding another unspecified-elsewhere invariant (the return layout of an exit must be the bottom layout).

**Pull request merged**
